### PR TITLE
Add RPC context handle wrapper

### PR DESCRIPTION
**Pull request by jonwis**
Here's an rpc-specific expansion of `wil::unique_any` that wraps a context handle and ensures the call to any `Close` method (which is itself an RPC call) is properly guarded against infrastructure failures.

### CODE REVIEW

File: include/wil/rpc_helpers.h [hunk 1]
```diff
@@ -132,6 +131,43 @@ namespace wil
         RpcEndExcept
     }
 
+    namespace details
+    {
+        // Provides an adapter around calling the context-handle-close method on an
+        // RPC interface, which itself is an RPC call.
+        template<typename TStorage, typename close_fn_t, close_fn_t close_fn>
```
Comment (oldnewthing) [inline include/wil/rpc_helpers.h]: Will the `TStorage` always be a trivially-copyable thing no bigger than a pointer?

File: include/wil/rpc_helpers.h [hunk 2]
```diff
@@ -132,6 +131,43 @@ namespace wil
...
+    // void CloseFoo([in, out] PFOO_CONTEXT);
+    using unique_foo_context = wil::unique_context_handle<FOO_CONTEXT, decltype(&CloseFoo), CloseFoo>;
+    unique_foo_context context;
+    RETURN_IF_FAILED(wil::invoke_rpc_nothrow(OpenFoo, m_binding.get(), context.put()));
+    RETURN_IF_FAILED(wil::invoke_rpc_nothrow(UseFoo, context.get()));
+    context.reset();
+    ~~~
+    */
+    template<typename TContext, typename close_fn_t, close_fn_t close_fn>
+    using unique_context_handle = unique_any<TContext, decltype(&details::rpc_closer_t<TContext, close_fn_t, close_fn>::Close), details::rpc_closer_t<TContext, close_fn_t, close_fn>::Close>;
```
Comment (oldnewthing) [inline include/wil/rpc_helpers.h]: Perhaps `unique_rpc_context_handle` would be clearer. While the term "context handle" is used primarily for RPC, it is not exclusively used in that sense.

File: include/wil/rpc_helpers.h (hunk 1 above)
Comment (jonwis) [reply include/wil/rpc_helpers.h]: Pretty sure, RPC handles must be pointer-sized.

File: include/wil/rpc_helpers.h (hunk 2 above)
Comment (jonwis) [reply include/wil/rpc_helpers.h]: Good idea, will fix

**Pull request merged**